### PR TITLE
chore: update federation demo to include countries entity with nested keys

### PR DESCRIPTION
**Pull request by clayne11**
Also included cycles in the hobbies graph.
 
The existing demo has no subgraph cycles. Once cycles are introduced, there are issues with the query planner.

## Motivation and Context
I have found issues with the query planner in a couple situations:
- if the query includes mutual recursion, the query planner will fail
- if you include an entity with a nested key selection set, there is weird behavior

I updated the tests to showcase the issues.

# Some pointers

When you have an entity with a nested keys, querying it has some odd behaviors.

1. Querying just the nested object from the countries subgraph without asking for its key works fine.
Query:
```graphql
query Working {
  employees {
    id
    details {
      location {
        language
      }
    }
  }
}
```

Result:
```
{
  "data": {
    "employees": [
      {
        "id": 1,
        "details": {
          "location": {
            "language": "German"
          }
        }
      },
      {
        "id": 2,
        "details": {
          "location": {
            "language": "German"
          }
        }
      },
      {
        "id": 3,
        "details": {
          "location": {
            "language": "English"
          }
        }
      },
      {
        "id": 4,
        "details": {
          "location": {
            "language": "German"
          }
        }
      },
      {
        "id": 5,
        "details": {
          "location": {
            "language": "Ukrainian"
          }
        }
      },
      {
        "id": 7,
        "details": {
          "location": {
            "language": "Hindi"
          }
        }
      },
      {
        "id": 8,
        "details": {
          "location": {
            "language": "Hindi"
          }
        }
      },
      {
        "id": 10,
        "details": {
          "location": {
            "language": "Dutch"
          }
        }
      },
      {
        "id": 11,
        "details": {
          "location": {
            "language": "German"
          }
        }
      },
      {
        "id": 12,
        "details": {
          "location": {
            "language": "English"
          }
        }
      }
    ]
  }
}
```

2. Asking for just the key of the nested key object is also fine.
Query:
```graphql
query Working {
  employees {
    id
    details {
      location {
        key {
          name
        }
      }
    }
  }
}
```

Result:
```
{
  "data": {
    "employees": [
      {
        "id": 1,
        "details": {
          "location": {
            "key": {
              "name": "Germany"
            }
          }
        }
      },
      {
        "id": 2,
        "details": {
          "location": {
            "key": {
              "name": "Germany"
            }
          }
        }
      },
      {
        "id": 3,
        "details": {
          "location": {
            "key": {
              "name": "America"
            }
          }
        }
      },
      {
        "id": 4,
        "details": {
          "location": {
            "key": {
              "name": "Germany"
            }
          }
        }
      },
      {
        "id": 5,
        "details": {
          "location": {
            "key": {
              "name": "Ukraine"
            }
          }
        }
      },
      {
        "id": 7,
        "details": {
          "location": {
            "key": {
              "name": "India"
            }
          }
        }
      },
      {
        "id": 8,
        "details": {
          "location": {
            "key": {
              "name": "India"
            }
          }
        }
      },
      {
        "id": 10,
        "details": {
          "location": {
            "key": {
              "name": "Netherlands"
            }
          }
        }
      },
      {
        "id": 11,
        "details": {
          "location": {
            "key": {
              "name": "Germany"
            }
          }
        }
      },
      {
        "id": 12,
        "details": {
          "location": {
            "key": {
              "name": "England"
            }
          }
        }
      }
    ]
  }
}
```

3. When you ask for both the key *and* a field from the countries subgraph, the key stops getting returned.
Query:
```graphql
query Working {
  employees {
    id
    details {
      location {
        key {
          name
        }
        language
      }
    }
  }
}
```

Result:
```
{
  "data": {
    "employees": [
      {
        "id": 1,
        "details": {
          "location": {
            "key": {},
            "language": "German"
          }
        }
      },
      {
        "id": 2,
        "details": {
          "location": {
            "key": {},
            "language": "German"
          }
        }
      },
      {
        "id": 3,
        "details": {
          "location": {
            "key": {},
            "language": "English"
          }
        }
      },
      {
        "id": 4,
        "details": {
          "location": {
            "key": {},
            "language": "German"
          }
        }
      },
      {
        "id": 5,
        "details": {
          "location": {
            "key": {},
            "language": "Ukrainian"
          }
        }
      },
      {
        "id": 7,
        "details": {
          "location": {
            "key": {},
            "language": "Hindi"
          }
        }
      },
      {
        "id": 8,
        "details": {
          "location": {
            "key": {},
            "language": "Hindi"
          }
        }
      },
      {
        "id": 10,
        "details": {
          "location": {
            "key": {},
            "language": "Dutch"
          }
        }
      },
      {
        "id": 11,
        "details": {
          "location": {
            "key": {},
            "language": "German"
          }
        }
      },
      {
        "id": 12,
        "details": {
          "location": {
            "key": {},
            "language": "English"
          }
        }
      }
    ]
  }
}
```

4. Now when you ask for just the language field but add another entity query to a different graph after it, that entity query doesn't work.

Query:
```graphql
query Working {
  employees {
    id
    details {
      location {
        language
      }
    }
    hobbies {
      __typename
    }
  }
}
```

Result:
```
{
  "data": {
    "employees": [
      {
        "id": 1,
        "details": {
          "location": {
            "language": "German"
          }
        },
        "hobbies": null
      },
      {
        "id": 2,
        "details": {
          "location": {
            "language": "German"
          }
        },
        "hobbies": null
      },
      {
        "id": 3,
        "details": {
          "location": {
            "language": "English"
          }
        },
        "hobbies": null
      },
      {
        "id": 4,
        "details": {
          "location": {
            "language": "German"
          }
        },
        "hobbies": null
      },
      {
        "id": 5,
        "details": {
          "location": {
            "language": "Ukrainian"
          }
        },
        "hobbies": null
      },
      {
        "id": 7,
        "details": {
          "location": {
            "language": "Hindi"
          }
        },
        "hobbies": null
      },
      {
        "id": 8,
        "details": {
          "location": {
            "language": "Hindi"
          }
        },
        "hobbies": null
      },
      {
        "id": 10,
        "details": {
          "location": {
            "language": "Dutch"
          }
        },
        "hobbies": null
      },
      {
        "id": 11,
        "details": {
          "location": {
            "language": "German"
          }
        },
        "hobbies": null
      },
      {
        "id": 12,
        "details": {
          "location": {
            "language": "English"
          }
        },
        "hobbies": null
      }
    ]
  }
}
```

5. When you move that same query physically above the location query, it starts working again.
Query:
```graphql
query Working {
  employees {
    id
    hobbies {
      __typename
    }
    details {
      location {
        language
      }
    }
  }
}
```

Result:
```
{
  "data": {
    "employees": [
      {
        "id": 1,
        "hobbies": [
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Gaming"
          },
          {
            "__typename": "Other"
          },
          {
            "__typename": "Programming"
          },
          {
            "__typename": "Travelling"
          }
        ],
        "details": {
          "location": {
            "language": "German"
          }
        }
      },
      {
        "id": 2,
        "hobbies": [
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Gaming"
          },
          {
            "__typename": "Programming"
          }
        ],
        "details": {
          "location": {
            "language": "German"
          }
        }
      },
      {
        "id": 3,
        "hobbies": [
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Other"
          },
          {
            "__typename": "Travelling"
          }
        ],
        "details": {
          "location": {
            "language": "English"
          }
        }
      },
      {
        "id": 4,
        "hobbies": [
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Flying"
          },
          {
            "__typename": "Travelling"
          }
        ],
        "details": {
          "location": {
            "language": "German"
          }
        }
      },
      {
        "id": 5,
        "hobbies": [
          {
            "__typename": "Other"
          },
          {
            "__typename": "Other"
          },
          {
            "__typename": "Other"
          }
        ],
        "details": {
          "location": {
            "language": "Ukrainian"
          }
        }
      },
      {
        "id": 7,
        "hobbies": [
          {
            "__typename": "Gaming"
          },
          {
            "__typename": "Other"
          }
        ],
        "details": {
          "location": {
            "language": "Hindi"
          }
        }
      },
      {
        "id": 8,
        "hobbies": [
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Gaming"
          },
          {
            "__typename": "Other"
          }
        ],
        "details": {
          "location": {
            "language": "Hindi"
          }
        }
      },
      {
        "id": 10,
        "hobbies": [
          {
            "__typename": "Programming"
          },
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Other"
          },
          {
            "__typename": "Travelling"
          }
        ],
        "details": {
          "location": {
            "language": "Dutch"
          }
        }
      },
      {
        "id": 11,
        "hobbies": [
          {
            "__typename": "Other"
          }
        ],
        "details": {
          "location": {
            "language": "German"
          }
        }
      },
      {
        "id": 12,
        "hobbies": [
          {
            "__typename": "Programming"
          },
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Gaming"
          },
          {
            "__typename": "Travelling"
          }
        ],
        "details": {
          "location": {
            "language": "English"
          }
        }
      }
    ]
  }
}
```

6. Finally, if you add the key back in to the nested key object, the second query starts working, however the nested key still isn't returned.

Query:
```graphql
query Working {
  employees {
    id
    details {
      location {
        key {
          name
        }
        language
      }
    }
    hobbies {
      __typename
    }
  }
}
```

Result:
```
{
  "data": {
    "employees": [
      {
        "id": 1,
        "details": {
          "location": {
            "key": {},
            "language": "German"
          }
        },
        "hobbies": [
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Gaming"
          },
          {
            "__typename": "Other"
          },
          {
            "__typename": "Programming"
          },
          {
            "__typename": "Travelling"
          }
        ]
      },
      {
        "id": 2,
        "details": {
          "location": {
            "key": {},
            "language": "German"
          }
        },
        "hobbies": [
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Gaming"
          },
          {
            "__typename": "Programming"
          }
        ]
      },
      {
        "id": 3,
        "details": {
          "location": {
            "key": {},
            "language": "English"
          }
        },
        "hobbies": [
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Other"
          },
          {
            "__typename": "Travelling"
          }
        ]
      },
      {
        "id": 4,
        "details": {
          "location": {
            "key": {},
            "language": "German"
          }
        },
        "hobbies": [
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Flying"
          },
          {
            "__typename": "Travelling"
          }
        ]
      },
      {
        "id": 5,
        "details": {
          "location": {
            "key": {},
            "language": "Ukrainian"
          }
        },
        "hobbies": [
          {
            "__typename": "Other"
          },
          {
            "__typename": "Other"
          },
          {
            "__typename": "Other"
          }
        ]
      },
      {
        "id": 7,
        "details": {
          "location": {
            "key": {},
            "language": "Hindi"
          }
        },
        "hobbies": [
          {
            "__typename": "Gaming"
          },
          {
            "__typename": "Other"
          }
        ]
      },
      {
        "id": 8,
        "details": {
          "location": {
            "key": {},
            "language": "Hindi"
          }
        },
        "hobbies": [
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Gaming"
          },
          {
            "__typename": "Other"
          }
        ]
      },
      {
        "id": 10,
        "details": {
          "location": {
            "key": {},
            "language": "Dutch"
          }
        },
        "hobbies": [
          {
            "__typename": "Programming"
          },
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Other"
          },
          {
            "__typename": "Travelling"
          }
        ]
      },
      {
        "id": 11,
        "details": {
          "location": {
            "key": {},
            "language": "German"
          }
        },
        "hobbies": [
          {
            "__typename": "Other"
          }
        ]
      },
      {
        "id": 12,
        "details": {
          "location": {
            "key": {},
            "language": "English"
          }
        },
        "hobbies": [
          {
            "__typename": "Programming"
          },
          {
            "__typename": "Exercise"
          },
          {
            "__typename": "Gaming"
          },
          {
            "__typename": "Travelling"
          }
        ]
      }
    ]
  }
}
```

### CODE REVIEW

**Comment by clayne11** [comment]
Once you checkout this PR, run the following commands:

1. In one terminal:
```
go mod download
go run ./demo/cmd/all
```

2. In another terminal:
```
ROUTER_CONFIG_PATH=./demo/config.json ENGINE_ENABLE_REQUEST_TRACING=true DEV_MODE=true go run ./router/cmd/router
```

**Comment by jensneuse** [comment]
We're investigating this. Currently we won't merge it as we're looking into it, but eventually we will.

**Comment by clayne11** [comment]
Sounds good, I didn't necessarily want it to be merged, just wanted to highlight the issues I'm seeing in a reproducible way.

**Comment by jensneuse** [comment]
> This PR was marked stale due to lack of activity. It will be closed in 14 days.

not stale
we're still working on fixing nested keys

File: demo/pkg/subgraphs/employees/subgraph/schema.graphqls [hunk 1]
```diff
@@ -55,33 +64,41 @@ enum OperationType {
 
 type Operator implements RoleType {
   departments: [Department!]!
-  operatorType: [OperationType!]!
   title: [String!]!
+  employees: [Employee!]! @goField(forceResolver: true)
+  operatorType: [OperationType!]!
+}
+
+type Details @key(fields: "id") {
```
Comment (devsergiy) [inline demo/pkg/subgraphs/employees/subgraph/schema.graphqls:72]: Could we not make Details an entity?

It illustrates sharable objects, sharable planning is quite different from regular entity planning

**Comment by devsergiy** [comment]
Hi @clayne11 
thanks a lot for your findings!

We have merged fixes for the issues you have found in this PR
Will you be able to try it out?
Otherwise, I will have to create a new branch based on this one

I have one question about `Details` type. Were there any composition errors when it is not an entity?

**Pull request closed without merging**